### PR TITLE
MKLDNN layout: Support for activation operator

### PR DESCRIPTION
**Pull request by mozga-intel**
Waiting for supports of the mkldnn’s layout.

Please have a look at this activation operator supported by the MKLDNN’s layout and assess if we can keep this design of the code.

This code uses the implementation of layout which is available in the last pull-request. Therefore some of the function in this code are not available in this pull-request, but are available here

This version of code can be merged into the main branch when the pull-request with layout is accepted.

The concept of splits of the long code was proposed by @luotao1

Pull-request is related to #11040

### CODE REVIEW

**Comment by mozga-intel** [comment]
@luotao1, @tensor-tang  The code is prepared to the code-review process.

**Comment by mozga-intel** [comment]
"For some kinds of element wise operations (namely ReLU with alpha parameter equals 0) dst and src can be interchangeable for the backward pass, which allows performing in-place forward even for training."

**Comment by mozga-intel** [comment]
We have the "out-of-place" implementation of code.

File: paddle/fluid/operators/activation_mkldnn_op.cc [hunk 1]
```diff
@@ -35,188 +39,260 @@ std::string gethash(const mkldnn::memory::dims &operand_dims,
...
+  auto p_src_layout = std::make_shared<memory::format>(src_format);
+  dev_ctx.SetBlob(key_src_layout, p_src_layout);
 
   auto p_fwd = std::static_pointer_cast<mkldnn::eltwise_forward>(
       dev_ctx.GetBlob(key_fwd));
 
-  // save input data to be referred in backward path
-  auto p_src_data = std::make_shared<const T *>(src_data);
-  dev_ctx.SetBlob(key_src_data, p_src_data);
+  std::shared_ptr<memory> dst_memory;
```
Comment (tensor-tang) [inline paddle/fluid/operators/activation_mkldnn_op.cc:127]: Please try to use unique_ptr next time.

File: paddle/fluid/operators/activation_op.cc [hunk 2]
```diff
@@ -82,6 +83,7 @@ class ActivationOp : public framework::OperatorWithKernel {
     ctx->ShareLoD("X", /*->*/ "Out");
   }
 
+ protected:
```
Comment (tensor-tang) [inline paddle/fluid/operators/activation_op.cc:86]: why protect them?

File: paddle/fluid/operators/activation_mkldnn_op.cc [hunk 3]
```diff
@@ -35,188 +39,260 @@ std::string gethash(const mkldnn::memory::dims &operand_dims,
...
+  diff_x->set_layout(DataLayout::kMKLDNN);
+  diff_x->set_format(GetMKLDNNFormat(*diff_src_memory));
 }
-}  // anonymous namespace
 
 template <typename T, mkldnn::algorithm algorithm>
 struct MKLDNNActivationFunc : public BaseActivationFunctor<T> {
-  template <typename ExecContext>
-  void operator()(const ExecContext &ctx) const {
+  void operator()(const framework::ExecutionContext &ctx) const {
```
Comment (tensor-tang) [inline paddle/fluid/operators/activation_mkldnn_op.cc:288]: why remove the typename?

File: paddle/fluid/operators/activation_mkldnn_op.cc (hunk 1 above)
Comment (mozga-intel) [reply paddle/fluid/operators/activation_mkldnn_op.cc:127]: Ok. I will change it, next time.

File: paddle/fluid/operators/activation_op.cc (hunk 2 above)
Comment (mozga-intel) [reply paddle/fluid/operators/activation_op.cc:86]: Everyplace, where we have the `GetExpectedKernelType `function we use the `protected` keyword. This the only one reason, why I used it.

File: paddle/fluid/operators/activation_mkldnn_op.cc (hunk 3 above)
Comment (mozga-intel) [reply paddle/fluid/operators/activation_mkldnn_op.cc:288]: Because, I did't need to have it - redundant code.

**Comment by mozga-intel** [comment]
@luotao1 Could you continue the checking of this code, please.

**Review by luotao1 — APPROVED** [review]
LGTM！Thanks very much！

**Pull request merged**